api/client/copy.go

```diff
@@ -0,0 +1,218 @@
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
+
+	response, statusCode, err := cli.call("GET", urlStr, nil, false)
+	if err != nil {
+		return stat, err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return stat, fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	err = json.NewDecoder(response).Decode(&stat)
+
+	return stat, err
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) error {
+	// Get an absolute destination path.
+	dstPath, err := resolveLocalPath(dstPath)
+	if err != nil {
+		return err
+	}
+
+	// Start by stat-ing the source path in the container.
+	srcStat, err := cli.statContainerPath(srcContainer, srcPath)
+	if err != nil {
+		// Some error, like container or resource does not
+		// exist, so we can't perform the copy operation.
+		return err
+	}
+
+	query := make(url.Values, 1)
+	query.Set("path", srcPath)
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, statusCode, err := cli.call("GET", urlStr, nil, false)
+	if err != nil {
+		return err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  srcStat.Mode.IsDir(),
+	}
+
+	return archive.CopyTo(response, srcInfo, dstPath)
+}
+
+func (cli *DockerCli) copyToContainer(srcPath, dstContainer, dstPath string) error {
+	// Get an absolute source path.
+	srcPath, err := resolveLocalPath(srcPath)
+	if err != nil {
+		return err
+	}
+
+	// Prepare source copy info.
+	srcInfo, err := archive.CopyInfoStatPath(srcPath, true)
+	if err != nil {
+		return err
+	}
+
+	// Prepare destination copy info by stat-ing the container path.
+	dstInfo := archive.CopyInfo{Path: dstPath}
+	dstStat, err := cli.statContainerPath(dstContainer, dstPath)
+	if err == nil {
+		dstInfo.Exists, dstInfo.IsDir = true, dstStat.Mode.IsDir()
+	} else if strings.Contains(err.Error(), archive.ErrNotDirectory.Error()) {
+		// The destination was asserted to be a directory but exists as a file.
+		return err
+	}
+	// Ignore any other error and assume that the parent directory of the
+	// destination path exists, in which case the copy may still succeed.
+
+	content, err := archive.TarResource(srcPath)
+	if err != nil {
+		return err
+	}
+	defer content.Close()
+
+	dstDir, copyContent, err := archive.PrepareArchiveCopy(content, srcInfo, dstInfo)
+	if err != nil {
+		return err
+	}
+
+	query := make(url.Values, 1)
+	query.Set("dstDir", dstDir)
+
+	urlStr := fmt.Sprintf("/containers/%s/extract-to-dir?%s", dstContainer, query.Encode())
```

Comment:
but how would we handle it if the content were a directory?

Also, I really, really, want to avoid writing a whole filesystem API. I think the interface should simply be:

- get an archive of the contents of some path in the container.
- extract an archive to some existing directory in the container.

I currently have the `stat-path` endpoint out of necessity, but if it could be folded into a HEAD request on the `archive-path` endpoint, that would be quite nice.